File: web/src/lib/decoded.ts
import { create } from "zustand";
import type {
  DecodedRecord,
  DecodedRecordOf,
  DecoderEvent,
  DecoderEventOf,
  DecoderKind,
  ServerEvent,
} from "./types";

export const RING_CAPACITY = 2000;

export const STATION_CAPACITY = 1000;

export const FLUSH_MS = 100;

export interface Station {
  kind: DecoderKind;
  id: string;
  event: DecoderEvent;
  lastSeen: number;
  freqHz: number;
  deviceSet: number;
  channel: number;
  frames: number;
}

export type StationOf<K extends DecoderKind> = Omit<Station, "kind" | "event"> & {
  kind: K;
  event: DecoderEventOf<K>;
};

type FramesByKind = { [K in DecoderKind]?: readonly DecodedRecordOf<K>[] };
type StationsByKind = { [K in DecoderKind]?: readonly StationOf<K>[] };

export interface DecodedState {
  frames: FramesByKind;
  stations: StationsByKind;
  lost: number;
  received: number;
  push: (record: DecodedRecord) => void;
  hydrate: (records: readonly DecodedRecord[]) => void;
  reportLost: (count: number) => void;
  observe: (event: ServerEvent) => void;
  flush: () => void;
  ageOut: (maxAgeMs: number, nowMs?: number) => void;
  dropFrames: (match: (record: DecodedRecord) => boolean) => number;
  clear: () => void;
}

const pending: DecodedRecord[] = [];
const stationIndex = new Map<DecoderKind, Map<string, Station>>();
let flushTimer: ReturnType<typeof setTimeout> | null = null;

const NO_FRAMES: readonly never[] = Object.freeze([]);
const NO_STATIONS: readonly never[] = Object.freeze([]);

export const useDecodedStore = create<DecodedState>((set) => ({
  frames: {},
  stations: {},
  lost: 0,
  received: 0,

  push: (record) => {
    pending.push(record);
    if (flushTimer === null) {
      flushTimer = setTimeout(publish, FLUSH_MS);
    }
  },

  hydrate: (records) => {
    const touched = new Set<DecoderKind>();
    for (const record of records) {
      if (mergeStation(record)) {
        touched.add(record.event.kind);
      }
    }
    if (touched.size === 0) {
      return;
    }
    set((state) => {
      const stations = { ...state.stations };
      for (const kind of touched) {
        assignStations(stations, kind, snapshotStations(kind));
      }
      return { stations };
    });
  },

  reportLost: (count) => set((state) => ({ lost: state.lost + count })),

  observe: (event) => {
    if (event.type === "Decoded") {
      useDecodedStore.getState().push(event.data);
    } else if (event.type === "DecodedBacklog") {
      useDecodedStore.getState().hydrate(event.data.records);
    } else if (event.type === "DecodedLost") {
      useDecodedStore.getState().reportLost(event.data.count);
    }
  },

  flush: publish,

  ageOut: (maxAgeMs, nowMs = Date.now()) => {
    const cutoff = nowMs - maxAgeMs;
    const expired: DecoderKind[] = [];
    for (const [kind, stations] of stationIndex) {
      let dropped = false;
      for (const [id, station] of stations) {
        if (station.lastSeen < cutoff) {
          stations.delete(id);
          dropped = true;
        }
      }
      if (dropped) {
        expired.push(kind);
      }
    }
    if (expired.length === 0) {
      return;
    }
    set((state) => {
      const stations = { ...state.stations };
      for (const kind of expired) {
        assignStations(stations, kind, snapshotStations(kind));
      }
      return { stations };
    });
  },

  dropFrames: (match) => {
    const staged = pending.filter((record) => !match(record));
    let dropped = pending.length - staged.length;
    pending.splice(0, pending.length, ...staged);

    const published = useDecodedStore.getState().frames;
    const frames = { ...published };
    let touched = false;
    for (const [kind, slice] of Object.entries(published) as [
      DecoderKind,
      readonly DecodedRecord[],
    ][]) {
      const kept = slice.filter((record) => !match(record));
      if (kept.length !== slice.length) {
        dropped += slice.length - kept.length;
        touched = true;
        assignFrames(frames, kind, kept);
      }
    }
    if (touched) {
      set({ frames });
    }
    return dropped;
  },

  clear: () => {
    pending.length = 0;
    stationIndex.clear();
    cancelFlush();
    set({ frames: {}, stations: {}, lost: 0, received: 0 });
  },
}));

export function useDecodedKind<K extends DecoderKind>(kind: K): readonly DecodedRecordOf<K>[] {
  return useDecodedStore((state) => state.frames[kind] ?? NO_FRAMES);
}

export function useStations<K extends DecoderKind>(kind: K): readonly StationOf<K>[] {
  return useDecodedStore((state) => state.stations[kind] ?? NO_STATIONS);
}

function publish(): void {
  cancelFlush();
  if (pending.length === 0) {
    return;
  }
  const batch = pending.splice(0, pending.length);
  const byKind = new Map<DecoderKind, DecodedRecord[]>();
  for (const record of batch) {
    const kind = record.event.kind;
    let group = byKind.get(kind);
    if (group === undefined) {
      group = [];
      byKind.set(kind, group);
    }
    group.push(record);
    mergeStation(record);
  }

  useDecodedStore.setState((state) => {
    const frames = { ...state.frames };
    const stations = { ...state.stations };
    for (const [kind, added] of byKind) {
      added.reverse();
      const next = added.concat(state.frames[kind] ?? NO_FRAMES);
      if (next.length > RING_CAPACITY) {
        next.length = RING_CAPACITY;
      }
      assignFrames(frames, kind, next);
      if (stationIndex.has(kind)) {
        assignStations(stations, kind, snapshotStations(kind));
      }
    }
    return { frames, stations, received: state.received + batch.length };
  });
}

function cancelFlush(): void {
  if (flushTimer !== null) {
    clearTimeout(flushTimer);
    flushTimer = null;
  }
}

function mergeStation(record: DecodedRecord): boolean {
  const id = stationId(record.event);
  if (id === null) {
    return false;
  }
  const kind = record.event.kind;
  let stations = stationIndex.get(kind);
  if (stations === undefined) {
    stations = new Map();
    stationIndex.set(kind, stations);
  }
  const previous = stations.get(id);
  stations.set(id, {
    kind,
    id,
    event: previous === undefined ? record.event : mergeForward(previous.event, record.event),
    lastSeen: recordTime(record),
    freqHz: record.freq_hz,
    deviceSet: record.device_set,
    channel: record.channel,
    frames: (previous?.frames ?? 0) + 1,
  });
  evictOldest(stations);
  return true;
}

function evictOldest(stations: Map<string, Station>): void {
  if (stations.size <= STATION_CAPACITY) {
    return;
  }
  const byAge = [...stations.entries()].sort((a, b) => a[1].lastSeen - b[1].lastSeen);
  for (const [id] of byAge.slice(0, stations.size - STATION_CAPACITY)) {
    stations.delete(id);
  }
}

function stationId(event: DecoderEvent): string | null {
  switch (event.kind) {
    case "adsb":
      return event.data.icao;
    case "ais":
      return String(event.data.mmsi);
    case "aprs":
      return event.data.source;
    case "pocsag":
      return String(event.data.address);
    case "flex":
      return String(event.data.address);
    case "ermes":
      return String(event.data.local_address);
    case "rds":
      return event.data.pi ?? null;
    case "navtex":
    case "acars":
    case "subghz":
    case "selcall":
    case "rtty":
    case "morse":
    case "cw_skimmer":
    case "psk31":
    case "psk63":
    case "ft8":
    case "ft4":
    case "wspr":
    case "tone":
    case "dv":
    case "ident":
    case "broadcast":
    case "radio_clock":
    case "gnss":
      return null;
  }
}

function mergeForward(previous: DecoderEvent, next: DecoderEvent): DecoderEvent {
  const data: Record<string, unknown> = { ...previous.data };
  for (const [key, value] of Object.entries(next.data)) {
    if (value != null) {
      data[key] = value;
    }
  }
  return { kind: next.kind, data } as DecoderEvent;
}

function recordTime(record: DecodedRecord): number {
  const at = Date.parse(record.at);
  return Number.isNaN(at) ? Date.now() : at;
}

function snapshotStations(kind: DecoderKind): readonly Station[] {
  const stations = stationIndex.get(kind);
  return stations === undefined ? NO_STATIONS : [...stations.values()];
}

function assignFrames(target: FramesByKind, kind: DecoderKind, value: readonly DecodedRecord[]) {
  (target as Record<string, readonly DecodedRecord[]>)[kind] = value;
}

function assignStations(target: StationsByKind, kind: DecoderKind, value: readonly Station[]) {
  (target as Record<string, readonly Station[]>)[kind] = value;
}

File: apps/native/src/decoded.rs
use std::{
    collections::{HashMap, VecDeque},
    sync::Arc,
};

use sdrmm_wire::decode::{DecodedRecord, DecoderEvent};
use serde_json::Value;

pub const RING_CAPACITY: usize = 2000;
pub const BROADCAST_DATA_CAPACITY: usize = 8;
pub const STATION_CAPACITY: usize = 1000;

pub type Frames = Arc<VecDeque<Arc<DecodedRecord>>>;
pub type Stations = Arc<HashMap<String, Arc<Station>>>;

#[derive(Clone, Debug, PartialEq)]
pub struct Station {
    pub kind: &'static str,
    pub id: String,
    pub event: DecoderEvent,
    pub last_seen_ms: i64,
    pub freq_hz: f64,
    pub device_set: u32,
    pub channel: u32,
    pub frames: u64,
}

#[derive(Clone, Debug, Default)]
pub struct Decoded {
    frames: HashMap<&'static str, Frames>,
    stations: HashMap<&'static str, Stations>,
    pub lost: u64,
    pub received: u64,
}

impl Decoded {
    #[must_use]
    pub fn frames(&self, kind: &str) -> Option<&Frames> {
        self.frames.get(kind)
    }

    #[must_use]
    pub fn stations(&self, kind: &str) -> Option<&Stations> {
        self.stations.get(kind)
    }

    pub fn all_frames(&self) -> impl Iterator<Item = &Arc<DecodedRecord>> {
        self.frames.values().flat_map(|slice| slice.iter())
    }

    pub fn publish(&mut self, batch: Vec<DecodedRecord>) {
        self.received += batch.len() as u64;
        for record in batch {
            self.merge_station(&record);
            let kind = record.event.kind();
            let slice = Arc::make_mut(self.frames.entry(kind).or_default());
            slice.push_front(Arc::new(record));
            slice.truncate(capacity_of(kind));
        }
    }

    pub fn hydrate(&mut self, records: &[DecodedRecord]) -> bool {
        let mut touched = false;
        for record in records {
            touched |= self.merge_station(record);
        }
        touched
    }

    pub fn report_lost(&mut self, count: u64) {
        self.lost += count;
    }

    #[must_use]
    pub fn stale(&self, max_age_ms: i64, now_ms: i64) -> bool {
        let cutoff = now_ms - max_age_ms;
        self.stations.values().any(|stations| {
            stations
                .values()
                .any(|station| station.last_seen_ms < cutoff)
        })
    }

    pub fn age_out(&mut self, max_age_ms: i64, now_ms: i64) {
        let cutoff = now_ms - max_age_ms;
        for stations in self.stations.values_mut() {
            if stations
                .values()
                .any(|station| station.last_seen_ms < cutoff)
            {
                Arc::make_mut(stations).retain(|_, station| station.last_seen_ms >= cutoff);
            }
        }
    }

    pub fn drop_frames(&mut self, matches: impl Fn(&DecodedRecord) -> bool) -> usize {
        let mut dropped = 0;
        for slice in self.frames.values_mut() {
            let hit = slice.iter().filter(|record| matches(record)).count();
            if hit > 0 {
                dropped += hit;
                Arc::make_mut(slice).retain(|record| !matches(record));
            }
        }
        dropped
    }

    fn merge_station(&mut self, record: &DecodedRecord) -> bool {
        let Some(id) = station_id(&record.event) else {
            return false;
        };
        let kind = record.event.kind();
        let stations = Arc::make_mut(self.stations.entry(kind).or_default());
        let previous = stations.get(&id);
        let event = previous.map_or_else(
            || record.event.clone(),
            |previous| merge_forward(&previous.event, &record.event),
        );
        let frames = previous.map_or(0, |previous| previous.frames) + 1;
        stations.insert(
            id.clone(),
            Arc::new(Station {
                kind,
                id,
                event,
                last_seen_ms: record_time_ms(record),
                freq_hz: record.freq_hz,
                device_set: record.device_set,
                channel: record.channel,
                frames,
            }),
        );
        evict_oldest(stations);
        true
    }
}

fn capacity_of(kind: &str) -> usize {
    if kind == "broadcast_data" {
        BROADCAST_DATA_CAPACITY
    } else {
        RING_CAPACITY
    }
}

fn evict_oldest(stations: &mut HashMap<String, Arc<Station>>) {
    let excess = stations.len().saturating_sub(STATION_CAPACITY);
    if excess == 0 {
        return;
    }
    let mut by_age: Vec<(i64, String)> = stations
        .values()
        .map(|station| (station.last_seen_ms, station.id.clone()))
        .collect();
    by_age.sort_unstable();
    for (_, id) in by_age.into_iter().take(excess) {
        stations.remove(&id);
    }
}

#[must_use]
pub fn station_id(event: &DecoderEvent) -> Option<String> {
    match event {
        DecoderEvent::Adsb(message) => Some(message.icao.clone()),
        DecoderEvent::Ais(message) => Some(message.mmsi.to_string()),
        DecoderEvent::Aprs(packet) => Some(packet.source.clone()),
        DecoderEvent::Call(call) => call.source.map(|source| source.to_string()),
        DecoderEvent::Pocsag(page) => Some(page.address.to_string()),
        DecoderEvent::Flex(page) => Some(page.address.to_string()),
        DecoderEvent::Ermes(page) => Some(page.local_address.to_string()),
        DecoderEvent::Rds(update) => update.pi.clone(),
        DecoderEvent::Dect(frame) => frame.identity.as_ref().map(|id| id.rfpi.clone()),
        DecoderEvent::Df(bearing) => bearing.station_id.clone(),
        _ => None,
    }
}

#[must_use]
pub fn merge_forward(previous: &DecoderEvent, next: &DecoderEvent) -> DecoderEvent {
    let (Ok(Value::Object(mut older)), Ok(Value::Object(newer))) =
        (serde_json::to_value(previous), serde_json::to_value(next))
    else {
        return next.clone();
    };
    if let (Some(Value::Object(data)), Some(Value::Object(fresh))) =
        (older.get_mut("data"), newer.get("data"))
    {
        for (key, value) in fresh {
            if !value.is_null() {
                data.insert(key.clone(), value.clone());
            }
        }
    } else {
        return next.clone();
    }
    serde_json::from_value(Value::Object(older)).unwrap_or_else(|_| next.clone())
}

#[must_use]
pub fn time_ms(at: &str) -> Option<i64> {
    at.parse::<jiff::Timestamp>()
        .ok()
        .map(jiff::Timestamp::as_millisecond)
}

#[must_use]
pub fn now_ms() -> i64 {
    jiff::Timestamp::now().as_millisecond()
}

fn record_time_ms(record: &DecodedRecord) -> i64 {
    time_ms(&record.at).unwrap_or_else(now_ms)
}

#[cfg(test)]
mod tests;
